mitosheet/mitosheet/code_chunks/code_chunk_utils.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# Copyright (c) Saga Inc.
+# Distributed under the terms of the GPL License.
+
+
+from copy import copy
+from typing import TYPE_CHECKING, List, Optional, Any
+
+from mitosheet.code_chunks.code_chunk import CodeChunk
+from mitosheet.code_chunks.step_performers.column_steps.delete_column_code_chunk import DeleteColumnsCodeChunk
+
+if TYPE_CHECKING:
+    from mitosheet.step import Step
+else:
+    Step = Any
+
+def optimize_code_chunks(all_code_chunks: List[CodeChunk]) -> List[CodeChunk]:
+    """
+    Given a list of code chunks, will recursively attempt to optimize them 
+    down to the smallest list of code chunks that have the same effects
+    as the original list. 
+
+    This is necessarily recursive, because of a situation like [A, A, B, B], 
+    where A and B can be combined to a No-op. Thus, after one call, we end with 
+    [A, B], and we need to recurse to finish the optimization.
+    """
+
+    all_code_chunks_reversed = copy(all_code_chunks)
+    all_code_chunks_reversed.reverse() # Reverse the list so we can pop from the old front quickly
+
+    code_chunks_list: List[CodeChunk] = []
+
+    optimized = False
+    while len(all_code_chunks_reversed) >= 2:
+        first_code_chunk = all_code_chunks_reversed.pop()
+        second_code_chunk = all_code_chunks_reversed.pop()
+
+        combined_chunk = first_code_chunk.combine_right(second_code_chunk)
+
+        if combined_chunk is not None:
+            # If we can combine the two chunks into one, take that
+            optimized = True
+            all_code_chunks_reversed.append(combined_chunk)
+        else:
+            # If we cannot combine the two chunks, take the first chunk, 
+            # and reset the second one to the new first chunk for the next loop
+            code_chunks_list.append(first_code_chunk)
+            all_code_chunks_reversed.append(second_code_chunk)            
+    
+    # Make sure we take the final item in the code chunks list, 
+    # as it has nothing ot combine_right with
+    if len(all_code_chunks_reversed) == 1:
+        code_chunks_list.append(all_code_chunks_reversed[-1])
```

Comment:
It seems simpler to me to do all_code_chunks_reversed[0] since what we're doing is checking that only one element exists, we can just take that one element. Its not about the element being the last one in the array, its that its the only one. But nbd to me